contrib/colab/Distributing_Large_Embedding_tables.ipynb

```diff
@@ -0,0 +1,495 @@
+        "    self.embedding_size = emb_size\n",
+        "    self._world_size = world_size\n",
+        "    self._distribute_embeddings()\n",
+        "\n",
+        "  def _distribute_embeddings(self):\n",
+        "    vocab_size = self.model.encoder.embed_tokens.weight.size(0)\n",
+        "    self.padding_idx = self.model.encoder.embed_tokens.padding_idx\n",
+        "    self.embedding = DistributedEmbedding(\n",
+        "        vocab_size, self.embedding_size, world_size=self._world_size)\n",
+        "    # We remove the original embedding layer.\n",
+        "    self.model.encoder.embed_tokens = None\n",
+        "\n",
+        "  def init_emb_weights(self):\n",
+        "    std = self.embedding_size\n",
+        "    nn.init.normal_(self.embedding.embeddings.weight, mean=0, std=std**-0.5)\n",
+        "    nn.init.constant_(self.embedding.embeddings.weight[self.padding_idx], 0)\n",
+        "\n",
+        "  def forward(self, **kwargs):\n",
+        "    inputs = kwargs['src_tokens']\n",
+        "    embedded_batch, emb_globalbatch_dimsliced = self.embedding(inputs)\n",
+        "    x = F.dropout(\n",
+        "        embedded_batch, p=self.dropout, training=self.model.training)\n",
+        "    # Hack the encoder's `forward_embedding` method so that it returns what\n",
+        "    #   was just computed in distributed fashion.\n",
+        "    # This needs to return two values.\n",
+        "    self.model.encoder.forward_embedding = lambda _: (x, None)\n",
+        "    return self.model(**kwargs), embedded_batch, emb_globalbatch_dimsliced\n",
+        "\n",
+        "  def emb_backward(self, *args, **kwargs):\n",
+        "    self.embedding.backward(*args, **kwargs)\n",
+        "\n",
+        "  def non_distr_params(self):\n",
+        "    # Last parameter is the added distributed embedding table.\n",
+        "    last_index = len(list(self.parameters())) - 1\n",
+        "    for i, _ in enumerate(self.parameters()):\n",
+        "      if i != last_index:\n",
+        "        yield _"
+      ],
+      "execution_count": 0,
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "5wahSapg6YAE",
+        "colab_type": "text"
+      },
+      "source": [
+        "Let's now create the `Namespace`, which `fairseq` uses to define the task, dataset, model and more."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "HCcPvjX36i_7",
+        "colab_type": "code",
+        "colab": {}
+      },
+      "source": [
+        "import argparse\n",
+        "\n",
+        "# The following  leads to an HBM OOM w/ the regular way of embedding tokens.\n",
+        "#   On v3-8:\n",
+        "EMBEDDING_SIZE = 4096\n",
+        "INPUT_SHAPES = [[64, 64],]\n",
+        "#   On v2-8:\n",
+        "EMBEDDING_SIZE = 2048\n",
+        "INPUT_SHAPES = [[64, 64],]\n",
+        "\n",
+        "args = argparse.Namespace(\n",
```

Comment:
Doesn't this huge blob have some configuration defaults?
This is more code than the distributed embedding class, which should kinda be the focus of the Colab 😉

Reply:
Let me look into trying to get this from fairseq's argument parser, but generally speaking, fairseq works with lots of command line arguments. I probably can get it down to 20-30 arguments, rest being defaults though.

BTW I folded this code block in colab, which looks much better than it does here, although I'm not sure if there's a way to do that by default here.